File: apps/web/src/types/Job.ts
export type RiskLevel = "High" | "Medium" | "Low";
export type Severity = "High" | "Medium" | "Low";
export type JobStatus =
  | "Saved"
  | "Applied"
  | "Screening"
  | "Interview"
  | "Offer"
  | "Rejected"
  | "Ghosted"
  | "No response";

export interface Finding {
  id: string;
  ruleId: string;
  severity: Severity;
  confidence: "High" | "Medium" | "Low";
  category: string;
  title: string;
  evidence: string;
  description: string;
  explanation: string;
  recommendation: string;
  scoreImpact: number;
  points: number;
}

export interface PositiveSignal {
  ruleId: string;
  title: string;
  evidence: string;
  description: string;
}

export interface PostingContext {
  postingDate?: string | null;
  firstSeen: string;
  mostRecentlySeen: string;
  observedAgeDays: number;
  repeatCount: number;
  possibleReposting: boolean;
}

export interface CategoryScores {
  phishing: number;
  fakeRecruiter: number;
  scam: number;
  ghost: number;
}

export interface Job {
  id: string;
  company: string;
  title: string;
  platform: string;
  date: string;
  riskLevel: RiskLevel;
  status: JobStatus;
  sourceUrl?: string | null;
  location?: string | null;
  overallScore?: number | null;
  scanDate?: string | null;
  dateApplied?: string | null;
  postingDate?: string | null;
  topFinding?: string | null;
  categories?: CategoryScores | null;
  findings?: Finding[] | null;
}

export type JobWithLatestReport = Job;

export interface JobUpdateRequest {
  company?: string;
  title?: string;
  platform?: string;
  sourceUrl?: string;
  location?: string | null;
  status?: JobStatus;
  dateApplied?: string | null;
}

File: backend/app/schemas/job.py
from datetime import date, datetime
from uuid import UUID

from pydantic import Field, model_validator

from app.schemas.common import ApiModel, JobStatus, RiskLevel, Severity


class Finding(ApiModel):
    id: str
    rule_id: str
    severity: Severity
    confidence: str
    category: str
    title: str
    evidence: str
    description: str
    explanation: str
    recommendation: str
    score_impact: int = Field(ge=0)
    points: int = Field(ge=0)

    @model_validator(mode="before")
    @classmethod
    def preserve_legacy_findings(cls, value):
        if not isinstance(value, dict):
            return value
        normalized = dict(value)
        normalized.setdefault("ruleId", normalized.get("rule_id", "LEGACY_FINDING"))
        normalized.setdefault("confidence", "Medium")
        explanation = normalized.get("explanation") or normalized.get("description", "")
        normalized.setdefault("explanation", explanation)
        normalized.setdefault("description", explanation)
        points = normalized.get("scoreImpact", normalized.get("score_impact", normalized.get("points", 0)))
        normalized.setdefault("scoreImpact", points)
        normalized.setdefault("points", points)
        return normalized


class PositiveSignal(ApiModel):
    rule_id: str
    title: str
    evidence: str
    description: str


class PostingContext(ApiModel):
    posting_date: date | None = None
    first_seen: datetime
    most_recently_seen: datetime
    observed_age_days: int = Field(ge=0)
    repeat_count: int = Field(ge=1)
    possible_reposting: bool = False


class CategoryScores(ApiModel):
    phishing: int = Field(default=0, ge=0, le=100)
    fake_recruiter: int = Field(default=0, ge=0, le=100)
    scam: int = Field(default=0, ge=0, le=100)
    ghost: int = Field(default=0, ge=0, le=100)


class JobCreate(ApiModel):
    company: str = Field(min_length=1, max_length=200)
    title: str = Field(min_length=1, max_length=250)
    platform: str = Field(min_length=1, max_length=100)
    source_url: str = Field(min_length=1, max_length=2048)
    location: str | None = Field(default=None, max_length=200)
    status: JobStatus = JobStatus.applied
    date_applied: date | None = None
    posting_date: date | None = None


class JobUpdate(ApiModel):
    company: str | None = Field(default=None, min_length=1, max_length=200)
    title: str | None = Field(default=None, min_length=1, max_length=250)
    platform: str | None = Field(default=None, min_length=1, max_length=100)
    source_url: str | None = Field(default=None, min_length=1, max_length=2048)
    location: str | None = Field(default=None, max_length=200)
    status: JobStatus | None = None
    date_applied: date | None = None


class JobRead(ApiModel):
    id: UUID
    company: str
    title: str
    platform: str
    date: str
    risk_level: RiskLevel
    status: JobStatus
    source_url: str | None = None
    location: str | None = None
    overall_score: int | None = Field(default=None, ge=0, le=100)
    scan_date: datetime | None = None
    date_applied: date | None = None
    top_finding: str | None = None
    categories: CategoryScores | None = None
    findings: list[Finding] | None = None


class JobWithLatestReport(JobRead):
    pass


Job = JobRead
